### PR TITLE
Strict common ethereum

### PR DESCRIPTION
**Pull request by yoozo**
# Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)
Enable TS strict model in common-ethereum package

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] I have tested locally
- [x] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [x] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] My code is up to date with the base branch
- [x] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/common-ethereum/src/codegen/codegen-controller.ts [hunk 1]
```diff
@@ -52,9 +52,19 @@ function validateAbi(datasources: SubqlRuntimeDatasource[], projectPath: string)
     const abiName = datasource.options.abi;
     const topicIssues: string[] = [];
     const funcIssues: string[] = [];
-    const abi = datasource.assets.get(abiName);
+
+    if (!datasource.assets) {
+      issues.push(`Databases.Asset is undefined`);
+      continue;
+    }
```
Comment (stwiname) [inline packages/common-ethereum/src/codegen/codegen-controller.ts:59]: I think line 63 can be `const abi = datasource.assets?.get(abiName);` This error isn't very useful

File: packages/common-ethereum/src/codegen/codegen-controller.ts [hunk 2]
```diff
@@ -133,31 +147,28 @@ export function prepareSortedAssets(
...
+    // We use actual abi file name instead on name provided in assets
+    // This is aligning with files in './ethers-contracts'
+    sortedAssets[parseContractPath(filePath).name] = value.file;
+  };
+
+  for (const d of datasources) {
+    if (!d.assets) continue;
+    if (!isRuntimeDs(d) && !isCustomDs(d) && !validateCustomDsDs(d)) continue;
+
+    for (const [name, value] of d.assets.entries()) {
```
Comment (stwiname) [inline packages/common-ethereum/src/codegen/codegen-controller.ts:167]: This has dropped support for the type being an object.

File: packages/types/src/project.ts [hunk 3]
```diff
@@ -258,7 +258,7 @@ export type SecondLayerHandlerProcessor<
   | SecondLayerHandlerProcessor_0_0_0<EthereumRuntimeHandlerInputMap, K, F, E, DS, ApiWrapper>
   | SecondLayerHandlerProcessor_1_0_0<EthereumRuntimeHandlerInputMap, K, F, E, DS, ApiWrapper>;
 
-type SecondLayerHandlerProcessorArray<
+export type SecondLayerHandlerProcessorArray<
```
Comment (stwiname) [inline packages/types/src/project.ts:261]: I think the fix from https://github.com/subquery/subql/pull/2469 also needs to be applied to this file. `EthereumRuntimeHandlerInputMap` -> `EthereumRuntimeFilterMap`

File: packages/common-ethereum/src/codegen/codegen-controller.ts (hunk 2 above)
Comment (yoozo) [reply packages/common-ethereum/src/codegen/codegen-controller.ts:167]: Previously, support for objects was likely implemented for compatibility purposes. The `assets` is type Map. And the function generateAbis has already ensured that all incorrect assets are converted to Map.

**Pull request merged**